### DISCUSSION TITLE
gradle manager detect no deps when 3 dep strings present in single configuration

### DISCUSSION THREAD

**Discussion by bh-tt** [renovatebot/renovate#30232] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Mend Renovate hosted app on github.com

### Please tell us more about your question or problem

I think I've found a bug with the gradle manager. When some dependencies are disabled and a gradle file contains multiple dependency strings in a single `implementation()` call (or any other configuration), Renovate disables all dependencies in this block if at least one of them matches the disable rule. I have a [reproducer](https://github.com/bh-tt/renovate-tt-gradle-reproducer) which showcases the problem.

With the following configuration, I'd expect renovate to not see the 2 org.junit.jupiter dependencies, but I still expect to see the jakarta.ws.rs dependency. The dependency dashboard issue also does not even show this dependency as part of the build.gradle.

renovate.json
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:recommended"
  ],
  "packageRules": [
        {
      "matchDatasources": ["maven"],
      "matchPackagePrefixes": ["org.junit.jupiter"],
      "groupName": "All upstream dependencies",
      "enabled": false,
      "description": "group upstreams more than 1 month old together"
    }

  ]
}
```

build.gradle:
```
plugins {
  id "java-library"
}


dependencies {
    testImplementation(
            'org.junit.jupiter:junit-jupiter-api:5.6.2',
            'org.junit.jupiter:junit-jupiter-params:5.6.3',
            'jakarta.ws.rs:jakarta.ws.rs-api:2.1.6'
    )
    api(
            'org.wiremock:wiremock:3.8.0'
    )
}
```

### Logs (if relevant)

<details><summary>Logs</summary>

```
DEBUG: Post-massage config
{
  "config": {
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
      "mergeConfidence:all-badges",
      "config:recommended"
    ],
    "packageRules": [
      {
        "matchDatasources": [
          "maven"
        ],
        "matchPackagePrefixes": [
          "org.junit.jupiter"
        ],
        "groupName": "All upstream dependencies",
        "enabled": false,
        "description": [
          "group upstreams more than 1 month old together"
        ]
      }
    ]
  }
}
.....

DEBUG: Matched 1 file(s) for manager gradle: build.gradle
DEBUG: manager extract durations (ms)
{
  "managers": {
    "gradle": 41
  }
}

DEBUG: Found gradle package files
DEBUG: Found 1 package file(s)
INFO: Dependency extraction complete
{
  "baseBranch": "main"
  "stats": {
    "managers": {
      "gradle": {
        "fileCount": 1,
        "depCount": 1
      }
    },
    "total": {
      "fileCount": 1,
      "depCount": 1
    }
  }
}

DEBUG: Looking up org.wiremock:wiremock in repository https://repo.maven.apache.org/maven2/
DEBUG: Found 25 new releases for org.wiremock:wiremock in repository https://repo.maven.apache.org/maven2/
DEBUG: hostRules: no authentication for repo.maven.apache.org
DEBUG: PackageFiles.add() - Package file saved for base branch
{
  "baseBranch": "main"
}

DEBUG: Package releases lookups complete
{
  "baseBranch": "main"
}

DEBUG: branchifyUpgrades
DEBUG: detectSemanticCommits()
DEBUG: getCommitMessages
DEBUG: semanticCommits: detected "unknown"
DEBUG: semanticCommits: disabled
DEBUG: 1 flattened updates found: org.wiremock:wiremock
DEBUG: Returning 1 branch(es)
DEBUG: config.repoIsOnboarded=true
DEBUG: packageFiles with updates
{
  "baseBranch": "main"
  "config": {
    "gradle": [
      {
        "packageFile": "build.gradle",
        "datasource": "maven",
        "deps": [
          {
            "depName": "org.wiremock:wiremock",
            "currentValue": "3.7.0",
            "managerData": {
              "fileReplacePosition": 292,
              "packageFile": "build.gradle"
            },
            "fileReplacePosition": 292,
            "datasource": "maven",
            "registryUrls": [],
            "depType": "dependencies",
            "updates": [
              {
                "bucket": "non-major",
                "newVersion": "3.8.0",
                "newValue": "3.8.0",
                "releaseTimestamp": "2024-06-28T17:11:20.000Z",
                "newMajor": 3,
                "newMinor": 8,
                "newPatch": 0,
                "updateType": "minor",
                "branchName": "renovate/org.wiremock-wiremock-3.x"
              }
            ],
            "packageName": "org.wiremock:wiremock",
            "versioning": "gradle",
            "warnings": [],
            "sourceUrl": "https://github.com/wiremock/wiremock",
            "registryUrl": "https://repo.maven.apache.org/maven2",
            "homepage": "http://wiremock.org",
            "packageScope": "org.wiremock",
            "currentVersion": "3.7.0",
            "currentVersionTimestamp": "2024-06-21T13:38:23.000Z",
            "isSingleVersion": true,
            "fixedVersion": "3.7.0"
          }
        ]
      }
    ]
  }
}

```

</details>

**Comment by RahulGautamSingh** [comment]
> When some dependencies are disabled and a gradle file contains multiple dependency strings in a single implementation() call (or any other configuration), Renovate disables all dependencies in this block if at least one of them matches the disable rule

I tested on you repro and found the the disabling of other deps is not required, it seems we do not extract from the `testImplementation` configration at all.

> (or any other configuration),

Are any other configrations not working? If so can you add them too.

Our Gradle parser is custom made, so there could be some cases it might not be handling. I am not familiar with the gradle parser, so lets wait for a maintainer's comment for a proper review.

**Reply by bh-tt** [reply]
I cannot reproduce that, see the `thisdoeswork.gradle` file I just added that has no `org.junit.jupiter` dependencies in it. Renovate seems to find that dependency just fine.

**Reply by bh-tt** [reply]
It's possible the problem is not in the gradle parser but rather somewhere after that, as renovate seems to combine all found dependencies in a single block and then disables them all if only a single dependency matches a disable rule. That would be a wrong assumption in the grouping code, that all dependencies in a single block all share the same characteristics.

**Reply by RahulGautamSingh** [reply]
I re-tested and:

1. one dep inside `testImplementation` works fine
2. two deps inside `testImplementation` works fine
3. three deps inside `testImplementation` breaks: no deps detected
2. four deps inside `testImplementation` works fine

🤯

**Reply by RahulGautamSingh** [reply]
In the logs also we can see the same:

```
INFO: Dependency extraction complete
{
  "baseBranch": "main"
  "stats": {
    "managers": {
      "gradle": {
        "fileCount": 1,
        "depCount": 1
      }
    },
    "total": {
      "fileCount": 1,
      "depCount": 1
    }
  }
}
```

It means only one dep was detected in the `build.gradle` file, which is before we perform any matching.
Could you update your title to match the issue? `gradle manager detect no deps when 3 dep strings present in single configuration`

**Reply by bh-tt** [reply]
I just checked, it also occurs with e.g. the `api` config, so it is probably all configurations with exactly 3 dependencies?

**Reply by RahulGautamSingh** [reply, marked as answer]
Created an issue: https://github.com/renovatebot/renovate/issues/30234

**Reply by Churro** [reply]
The Gradle manager works agnostic to configurations. In this case, the issue is with heuristic matching: instead of detecting 3 independent deps, precedence is given to a group-name-version pattern. Group and name values are interpolated, so renovate figures out that the first and second values are full deps, whereas the third is interpreted as version value. Since a version value can't contain : it is discarded.

I may find time to look at this when I'm back from vacation.
